src/Compilers/Core/Portable/Diagnostic/CustomObsoleteDiagnosticInfo.cs

```diff
@@ -0,0 +1,108 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal sealed class CustomObsoleteDiagnosticInfo : DiagnosticInfo
+    {
+        internal ObsoleteAttributeData Data { get; }
+
+        internal CustomObsoleteDiagnosticInfo(CommonMessageProvider messageProvider, int errorCode, ObsoleteAttributeData data, params object[] arguments)
+            : base(messageProvider, errorCode, arguments)
+        {
+            Data = data;
+        }
+
+        public override string MessageIdentifier
+        {
+            get
+            {
+                var id = Data.DiagnosticId;
+                if (!string.IsNullOrEmpty(id))
+                {
+                    return id;
+                }
+
+                return base.MessageIdentifier;
+            }
+        }
+
+        private DiagnosticDescriptor? _descriptor;
```

Comment:
Consider grouping all fields / auto props at the top of the type.

Reply:
Done